### ISSUE TITLE
NuGet pack should not fail for release packages with prerelease development dependency

### ISSUE THREAD

**Issue by lostmsu** [NuGet/Home#1365]
Currently, release package can not have prerelease development dependency. For example, that prevents us from using StyleCop.Analyzers package currently in beta in any release package development,

Internal alias: vimilova

**Comment by yishaigalatzer** [comment]
That's currently by design. This prevents shipping release packages with pre-release dependencies accidentally.

We can consider adding a flag to override this behavior (and if you want to provide a PR for that we will gladly take it). For now though, I don't see this popping up to the top of our priority list.

Your workaround would be to edit the nuspec after the package was packed, or create a pre-release package with the analyzer.

In general its a pretty bad idea to include an analyzer with your implementation package anyways. You want that in a top level "meta" package that will bring both the implementation and the analyzer, now that package can totally be pre-release.

**Comment by zidad** [comment]
I'd be interested to see a way to override this. We internally we use nuget, and not being able to create release packages with beta dependencies is holding us back at the moment.

**Comment by GeertvanHorrik** [comment]
I would need something like this as well (since we use NuGet to ship extensions, we don't care about the prereleases). An option (like -allowprereleasedependencies) to override  would be great.

**Comment by asbjornu** [comment]
I find it very strange that when upgrading a package with `-Pre`, that flag isn't also cascaded to dependencies. It's very natural for prerelease packages to have prerelease dependencies, so I don't really understand the reasoning here. Although this issue is for packaging, I expect it to be related to the upgrade process, no?

The "by accident" argument is in my opinion only valid if the package is stable. If the package is prerelease, it should by definition be able to have prerelease dependencies without any upgrades to it breaking. As it stands now, you need to do `-IgnoreDep` alongside `-Pre` if the package you upgrade to have prerelease dependencies. Excuse my french, but that's just stupid.

**Comment by NikGovorov** [comment]
I have same issue, it would be perfect to have the flag that would allow to override default behaviour.

**Comment by peterlai-roboops** [comment]
This is pretty bad design, for example we name our version with the branch it came from, so it makes tracking easier. This is deemed a pre-release version as it has letters in it, and this means we can't package another with an internal package.

**Comment by FransBouma** [comment]
I also ran into this this morning. I have 1 main package and several depending packages. All depending packages reference the main package, so take a dependency on it. As everything is now in prerelease I can't reference the depending packages as nuget can't find the main package, because I use as dependency version range: `(>= 5.1.0 && < 5.2.0)`, and the main package is in prerelease and has as version 5.1.0-beta-20161027. There's no way to specify this! I really don't want to alter the nuspec files to include some sort of beta version number because that is horrible to automate as these packages are generated by a script. 

If A and B are in prerelease, and B depends on A, and I want to add the package B to my project and I specify '-pre', nuget should be able to get A, using the normal versions and append the prerelease suffix, i.e. should really look for 5.1.0.\* as it's a prerelease ! I now have no idea how to solve this. 

Thanks @asbjornu  for pointing out -ignoredep, that at least made me able to install the package.

**Comment by yishaigalatzer** [comment]
Hey Fran's,

Is your issue that you want pack to just allow the pre-release in the nuspec OR you want resolution to work even though you specified a stable dependency?

Yishai

**Comment by peterlai-roboops** [comment]
To be honest, it doesn't really matter what the issue is because there's are so many reason one would name their packages which falls foul of the "release" naming standard, that there really should be a flag to allow us to override this behaviour. 

And it's a pretty big behaviour, whether to be allowed to package our code or not.

**Comment by FransBouma** [comment]
@yishaigalatzer 

I have A and B, both prerelease, B depends on A. They're generated from scripts which get the version to work with passed to them. Therefore the nuget packages have '-beta-xxxx' as suffix in their filename, but not in their versions. So nuget packages for A and B are: 

sd.xxx.A.5.1.0-beta-xxx.dll
sd.xxx.B.5.1.0-beta-xxx.dll

but internally the version is set to 5.1.0, and dependencies of B, as it depends on A, is set to `(>=5.1.0 && < 5.2.0)`. 

If I now install B in my project, and specify -pre, I'll get the prerelease. I want nuget to then also consider the packages B depends on as prerelease, so it can _find_ them. In my scenario above, installing B prerelease, would then work as it would find sd.xxx.A.5.1.0-beta-xxx as that's a prerelease too and the version matches the version specification of the dependency in B. 

Today however, I can't install B at all unless I specify -ignoredep, as nuget searches for the release package 5.1.0 of A, even though B is prerelease, because the dependencies in the nuspec of B refer to `>=5.1.0` and for nuget 5.1.0-beta-xxx is not matching that criteria.

**Comment by yishaigalatzer** [comment]
That's because 5.1.0-betaXXX is indeed lower than 5.1.0. Try specifying 5.1.0-beta as the min dependency.

**Comment by FransBouma** [comment]
Yes but that's not what I want as that would mean I have to update the nuspec files every time I build a package, so it's not really automate friendly. This is especially unfriendly for packages which have multiple builds per alpha-beta cycle.

**Comment by yishaigalatzer** [comment]
That's not what I'm suggesting.

1.0.0-beta < 1.0.0-beta1 < 1.0.0

So specify 1.0.0-beta verbatim.

Note that 
A. We cannot make a change to the resolution here, it's a massive breaking change.
B. This is not the same issue the bug is discussing, so if you do want to keep debating this issue, let's take it somewhere else

**Comment by FransBouma** [comment]
Sorry, but how is what I ran into not the same as this issue? It's a side effect of the prerelease factor not being used for resolvement. If prerelease is seen as prerelease across the resolution of dependencies, it would search for prerelease packages too, which is IMHO the issue discussed here. If not, then this whole system is too convoluted to begin with as I have no idea what on earth it _is_ discussing. 

I also disagree with your 'massive breaking change' as it currently simply _doesn't work_. breaking something that doesn't work today isn't breaking anything. 

I'm sorry I'm a bit harsh on this, but I can't believe I have to hand-edit nuget spec files now just for alpha-beta releases and remember to roll back these changes when RTM hits as the automated system of course simply rolls on and doesn't care about this. 

There's NO documentation on this either. I searched, I ran into some SO questions, with vague references to old issues but that's it.

**Comment by yishaigalatzer** [comment]
1. This bug is about the ability to pack a release package with a pre-release dependency, not about resolution.
2. I gave you a method that is not requiring editing of dependencies
3. Changing resolution rules IS a breaking change, and we are not going to consider taking one here.
4. We can clarify the versioning order in docs, it will happen shortly.
5. There are numerous ways to automate creating the correct dependencies, and they should change if your dependencies change. In this case it will mean the nuspec in the end package should change between pre-release and release, it could change as you rev your pre-release depenednecies (more on that a later), you could keep it constant by requiring a static 1.0.0-alpha or 1.0.0-beta as a minimum as it will match the range.
6. Back to revving the min Version - it's a choice you need to make. In my projects I believe in revving it and in same cases making it strict to avoid collisions from refactoring and other breaking changes that happen a lot more often during pre-release. The policy here is something you should determine because you own the development process of these depenednecies.

**Comment by peterlai-roboops** [comment]
but how a package is determine to be pre-release isn't that good. If it has a word in it then that's it, it's a pre-release. 

Your basically determining how every company should name their package via this one feature.

**Comment by yishaigalatzer** [comment]
Yes we are determining what versions look like, that's what standardization is for, it makes everyone understand the same concept and enables tools to operate on a common ground.

We support semver 1.0.0 and just released initial support for semver 2.0.0 in NuGet 3.5.

Http://www.semver.org

**Comment by peterlai-roboops** [comment]
yes you are! but i hardly think its for you to determine how we want to use and build internal packages that we have no intention of sharing. 

If there are genuine reasons to change it, e.g. some new tools that requires it and if we think it's worth the adaptation then we might change our mind and adapt to it, but it's not for you to decide how our company should name our packages.

**Comment by FransBouma** [comment]
@yishaigalatzer Fair enough. I understood it is about the same issue as the mechanics are really the same: how to determine which packages match the criteria. Whether the starting package is prerelease and wants to refer to prerelease dependents too or release package wants to refer to prerelease or even prerelease which wants to solely refer to release packages, that's just semantics on the same feature: you have an assembly B which refers to A and B as certain characteristics and A as another set of certain characteristics, and the rules which packages match the criteria set by B's characteristics + dependency specifications so A does or doesn't match is currently too strict. Be it release B referring to prerelease A, prerelease B referring to prerelease A or prerelease B referring to release A (which is automatic), that doesn't matter: they're a set of rules. Currently they only work with hard-coded version numbers, they can't deal with _characteristics_ of the packages: if a package is in 'prerelease' that's something the rules can't deal with. If they did, then the issue at hand, be it release->prerelease, or prerelease->prerelease would not be a problem. 

This isn't really 'changing' the resolution rules, it's making them more flexible if they have to be. They're currently hard-coded and can't be changed. If they would have been more flexible, the OP would be able to specify what was required, and people like me would have been able to do that too, with the same change. Additionally, others out there who don't care about this aren't affected. 

I must say I hate the 'solution' to alter the nugetspec files with a hardcoded 'beta' reference in the dependency, especially because it's something in a file I have to take care of before RTM so change the file manually before the automated system will spit out a set of assets (we're not talking 1 package here, but many), _and_ because they're all versioned together, so B refers to A, and they're versioned together. To me a pretty default concept, but it's not possible to use this, other than hard-coded version references. As it's not documented either, it tells me the nuget team hasn't thought about any of this as something that people would do.

**Comment by asbjornu** [comment]
> I understood it is about the same issue as the mechanics are really the same: how to determine which packages match the criteria. Whether the starting package is prerelease and wants to refer to prerelease dependents too or release package wants to refer to prerelease or even prerelease which wants to solely refer to release packages, that's just semantics on the same feature

I completely agree with all of the above. 👍 

> I must say I hate the 'solution' to alter the nugetspec files with a hardcoded 'beta' reference in the dependency, especially because it's something in a file I have to take care of before RTM so change the file manually before the automated system will spit out a set of assets (we're not talking 1 package here, but many), and because they're all versioned together, so B refers to A, and they're versioned together.

Yep. We have many solutions with projects being packaged up as `.nupkg` where each project in the solution depends on another, all up to a `.Core` project at the top. When these projects are published as pre-release packages, they are basically impossible to install without adding `-Pre -IgnoreDep` when upgrading them.

To fix this according to @yishaigalatzer's description, it will require a lot of monkeypatching on each and every package to fix the dependency chain and remove any version number after the pre-release label.

**Comment by yishaigalatzer** [comment]
No one said editing the nuspec is ideal, I'm saying the nuspec inside the package has the right representation. Definitely a pain point that needs a discussion, on a separate issue.

The breaking change is very real, the pre-release version will install instead of the rtm one if it stays on the feed, when installing with minimal depenednecies mode (default).

**Comment by FransBouma** [comment]
> The breaking change is very real, the pre-release version will install instead of the rtm one if it stays on the feed, when installing with minimal depenednecies mode (default).

No, as what I suggested (and which solves the point in the OP as well) is a set of rule parameters which configure what Nuget should do in several situations. The current behavior can stay as-is. With additional parameters nuget can behave a little different, i.e. the rules get additional input / different input parameters and can with that decide whether a package P does or doesn't fall into the set of packages matching the criteria. This doesn't break anything as the current behavior doesn't change. It only offers additional behaviors based on input parameters for the rules (which are currently strict and not configurable to solve the points addressed in this thread).

**Comment by yishaigalatzer** [comment]
I beg you, start a new thread and jot out your suggestions in concrete terms

**Comment by FransBouma** [comment]
Why another thread for a solution that covers this thread too, as the problem of this thread is actually caused by the same cause as the problem I described, namely: rules which are too strict for a given situation?

**Comment by asbjornu** [comment]
@yishaigalatzer: The problem we're discussing does not seem to be  isolated to what this issue describes in its title, so can't the title of this issue just be edited instead? The problem is the dependency resolving algorithm, which is behaving the same way for both `nuget pack` and `nuget update / install`.

**Comment by FransBouma** [comment]
**EDIT**
The hotfix argument I made isn't really valid, as one can specify 'highest path' filter option with nuget and get a matching dependent package. For beta and custom versioning this isn't going to fly of course.

**Comment by yishaigalatzer** [comment]
Frans, I agree hand editing is awful. Brain storming ideas, will post when I have a good one

Sent from Ninehttp://www.9folders.com/

**Comment by peterlai-roboops** [comment]
just give us the option to override the behaviour. 

We should be able to decide what's best for us, if what we decide doesn't work with some other tool then that's our own fault and we will correct that, but you shouldn't be forcing it on us.

I just don't get why there's push back on this, the users are saying we 100% need this but for some reason you know best. Generally, what happens is someone else release a product that does do what the user want and we all move to that. e.g. maven, years we been telling them it's too rigid, and now gradle and sbt is taking over.

**Comment by haf** [comment]
@yishaigalatzer Can we override it yet? This is _real_ pain for me when I have a single target in Logary that depends on a pre-release package but the suite of targets as such isn't a pre-release. (Also, the idea that you can't transitively depend on pre-release packages and be non-prerelease is wrong)

**Comment by enricosada** [comment]
Seems i got lots of good company there (hi @haf).

I have the same problem, and is really annoying. 
It's just about `dotnet pack`, not resolution on restore.

Make it a warning if you really want, but not an error, you cannot know why i am using a prerelease deps.
Or add an argument to override or make it a warning (noisy but still).

**Comment by mythz** [comment]
Just got hit with this issue, I'm trying to build a .NET Core MySql package with `nuget pack` where the only [.NET Core package of MySql.Data available](https://www.nuget.org/packages/MySql.Data/7.0.6-IR31) is a pre-release package.

Things I tried to get it to build a NuGet package with a -pre release package dependency:

```xml
<dependency id="MySql.Data" version="[7.0.6-ir31, )" />
<dependency id="MySql.Data" version="7.0.6-ir31" />
<dependency id="MySql.Data" version="[7.0.6-ir31]" />
<dependency id="MySql.Data" version="[7.0.6-ir31]" developmentDependency="true" />
```

What worked:

None of them. We had to change our CI build itself to run a special-cased custom script **for just this one NuGet package** which looks like:

    nuget pack ServiceStack.OrmLite.MySql.Core.nuspec -version 1.0.%version%-pre

> hopefully this helps someone else as this workaround was not obvious

IMO they all should work, I'm clearly asking for the pre-release dependency. If I happen to accidentally publish a package with a pre-release dependency that would clearly be my fault for making a **hard-coded reference to a pre-release package**. In which case the fix would be the same as every other bad NuGet package, i.e. delist it and publish a new version. I personally never publish pre-release packages and thanks to running into unintuitive non-standard behavior like this, I never will.

**Comment by enricosada** [comment]
@yishaigalatzer @rrelyea any update on this? Please is possible to add a workaround to unblock us? Like:

- an env var `NUGET_SUPPRESS_ERR_1234` (temporary fix but easy without impact on anyone and no api/cli changes) until a real solution is choosen 
- downgrade the error as warning. 


I'll implement anyone of that and send PR if is approved. Or just tell what you want. 

This issue is really annoying and common becuase lots of netstandard lib are prerelease atm but is ok for my lib, because as author I know that. 

For f# lib is a big issue too becuase

1. Was not like that in preview2
2. The mandatory microsoft fsharp.core package is prerelease probably until VS2017 RTM

**Comment by yishaigalatzer** [comment]
We should be able to take a pr for this. I agree we should prioritize it otherwise

**Comment by enricosada** [comment]
@yishaigalatzer i'll send a PR tomorrow. 
What do you prefer? Downgrade as warning or env var or something else?
Ihmo preferred is downgrade as warning

**Comment by yishaigalatzer** [comment]
Yes but with a way to clear the warning or warning as error

Sent from Nine<http://www.9folders.com/>
________________________________

**Comment by lostmsu** [comment]
I think it should be a warning.

**Comment by peterlai-roboops** [comment]
OMG, was that all that was needed to advance this? we do the work and you approve? Its fantastic that this is being advanced but  I kinda wish this was said months ago.

**Comment by enricosada** [comment]
didnt forget about it, i'll send a PR max this weekend

**Comment by Joe4evr** [comment]
@enricosada So where's that PR now?

**Comment by rohit21agrawal** [comment]
This has already been fixed in the latest nightly builds. The error has been downgraded to a warning

https://github.com/NuGet/NuGet.Client/pull/1073

**Comment by BrainCrumbz** [comment]
Is that nuget.exe available through some channel to be downloaded, especially for using it from a CI environment? Thanks

**Comment by haf** [comment]
Wow, amazing turn-around time there!

**Comment by pantonis** [comment]
I have the same problem on my CI. I am using nuget 3.5.0. It is a pain. What version of nuget includes this fix?

**Comment by shytikov** [comment]
Ok, wait, someone uses this functionality on production and actually expects things to fail in case if pre-release dependencies exists.

I believe that behavior should be configurable. Now what is done is just breaking behavior with no workaround. How we supposed verify that package is allowed to be called stable?

**Comment by drypa** [comment]
stylecop analyzer should not go to *.nupkg
for me helps in .csproj file set for stylecop:
`<PackageReference Include="StyleCop.Analyzers" Version="1.1.0-beta008">
      <ExcludeAssets>all</ExcludeAssets>
    </PackageReference>`

**Comment by IanKemp** [comment]
For anyone still stumbling across this in 2022, the magic flag is `-AllowPreReleaseDependencies`. It doesn't seem to be documented anywhere.

**Comment by johnabroad42** [comment]
Hi i don't understand  about the  multiple  life

**Comment by johnabroad42** [comment]
Since 2018 Until now  always i talk  but no body fixed

**Comment by johnabroad42** [comment]
Every thing i report to fixed and help because  they dont want to help

**Comment by ardalis** [comment]
@IanKemp No such switch for nuget or for dotnet.

![image](https://user-images.githubusercontent.com/782127/199578884-32a4c821-f1d2-48c7-bb68-f5d9449d6e8b.png)

**Comment by zivkan** [comment]
@ardalis that command line argument only works for `nuget.exe pack`, which only supports packages.config.  For projects using PackageReference (and therefore msbuild pack), you can use NoWarn NU5104:

```xml
<PropertyGroup>
  <NoWarn>$(NoWarn);NU5104</NoWarn>
</PropertyGroup>
```

**Comment by tonyqus** [comment]
Looks I encounter the issue today. Here is the story from NPOI team. 

MSFT suggests the community to get rid of System.Drawing.Common since they don't wanna maintain this library anymore (There are a few weird excuse like libgdi is not well maintained. But why MSFT cannot hire someone to maintain it?). Then we have no choice but migrate to a new drawing package.  We finally choose ImageSharp. And then it takes about a month to migrate from System.Drawing.Common to SixLabors.Fonts. 

We released NPOI 2.6.0-rc3 (which reference SixLabors.Fonts) 2 month ago and everything looks fine (no errors). Then NPOI 2.6.0 is released today. It's almost the same nuget package from 2.6.0-rc3 except the version. However, Visual Studio cannot find SixLabors.Fonts in any case when the developer try to reference 2.6.0 package because nuget doesn't support dependency on a prerelease package. And it's interesting that a prerelease version can work well if it references a prerelease package.

I also tested ClosedXml package because it also reference SixLabors.Fonts package. Same error: 'Cannot resolve dependency SixLabors.Fonts'. 

========Update ========
I can reproduce it  while using .NET framework 4.7.2 project and it's a project using packages.config file

**Comment by zivkan** [comment]
@tonyqus 

> nuget doesn't support dependency on a prerelease package.

Yes it does. For years, NuGet treats it as a warning that can be ignored if you choose. See [my comment](https://github.com/NuGet/Home/issues/1365#issuecomment-1302696473) immediately before yours. 

In fact, [NPOI 2.6.0](https://www.nuget.org/packages/NPOI/#dependencies-body-tab) (a semver stable version) lists a dependency on SixLabors.Fonts version 1.0.0-beta18, so I don't understand why you're saying it's not supported when your package already does it. 

> Same error: 'Cannot resolve dependency SixLabors.Fonts'.

```text
D:\src\test\npoitest> dotnet new console
The template "Console App" was created successfully.

Processing post-creation actions...
Restoring D:\src\test\npoitest\npoitest.csproj:
  Determining projects to restore...
  Restored D:\src\test\npoitest\npoitest.csproj (in 68 ms).
Restore succeeded.

D:\src\test\npoitest> dotnet add package npoi
  Determining projects to restore...
  Writing C:\Users\zivkan\AppData\Local\Temp\tmp2B6A.tmp
info : X.509 certificate chain validation will use the default trust store selected by .NET.
info : X.509 certificate chain validation will use the default trust store selected by .NET.
info : Adding PackageReference for package 'npoi' into project 'D:\src\test\npoitest\npoitest.csproj'.
info :   GET https://api.nuget.org/v3/registration5-gz-semver2/npoi/index.json
info :   OK https://api.nuget.org/v3/registration5-gz-semver2/npoi/index.json 826ms
info : Restoring packages for D:\src\test\npoitest\npoitest.csproj...

// cutting out a bunch of lines
info :   GET https://api.nuget.org/v3-flatcontainer/sixlabors.fonts/index.json
info :   OK https://api.nuget.org/v3-flatcontainer/sixlabors.fonts/index.json 904ms
info :   GET https://api.nuget.org/v3-flatcontainer/sixlabors.fonts/1.0.0-beta18/sixlabors.fonts.1.0.0-beta18.nupkg
info :   OK https://api.nuget.org/v3-flatcontainer/sixlabors.fonts/1.0.0-beta18/sixlabors.fonts.1.0.0-beta18.nupkg 72ms
info : Installed SixLabors.Fonts 1.0.0-beta18 from https://api.nuget.org/v3/index.json with content hash evykNmy/kEE9EAEKgZm3MNUYXuMHFfmcLUNPw7Ho5q7OI96GFkkIxBm+QaKOTPBKw+L0AjKOs+ArVg8P40Ac9g==.
info : Installed NPOI 2.6.0 from https://api.nuget.org/v3/index.json with content hash Pwjo65CUH3MiRnBEbVo8ff31ZrDGdGyyFJyAEncmbTQ0/gYgDkBUnGKm20aLpdwCpPNLzvapZm8v5tx4S6qAWg==.
info : Package 'npoi' is compatible with all the specified frameworks in project 'D:\src\test\npoitest\npoitest.csproj'.info : PackageReference for package 'npoi' version '2.6.0' added to file 'D:\src\test\npoitest\npoitest.csproj'.
info : Writing assets file to disk. Path: D:\src\test\npoitest\obj\project.assets.json
log  : Restored D:\src\test\npoitest\npoitest.csproj (in 12.38 sec).
```

By any chance are you using a custom nuget feed, with nuget.org removed from your nuget.config?  It sounds to me like the sixlabors.fonts package hasn't been mirrored to your custom feed.

**Comment by tonyqus** [comment]
@zivkan No, I don't use any custom nuget feed. It's just nuget.org. The difference is that I don't use dotnet command in this case. I use Visual Studio - Add nuget reference. You can reproduce it easily if your project uses .NET framework 4.7.2. I tried both VS2019 and VS2022. 

Your console project looks to be .NET core by default instead of .NET framework 4.7.2. And My project is using packages.config 

I see ClosedXML provides [some workaround](https://github.com/ClosedXML/ClosedXML/blob/35a2fca7852a9a941287f6d5f828b10de0fd93ce/docs/installation.rst). This looks to be a problem with packages.config instead of .NET version.

**Comment by zivkan** [comment]
Indeed, PackageReference still works, even with the legacy csproj. I confirm that it failed with packages.config, but this thread is about packing, not installing packages. I can't find a duplicate issue from a quick search, so I suggest filing a new issue for that to be investigated. The topic of this issue has been resolved, and I believe this issue is correctly closed.

Since before 2016 NuGet wouldn't allow semver stables packages to depend on prerelease packages at all, this wasn't a scenario that was possible for the first 5 years that NuGet was around. Looks like it wasn't taken into account when https://github.com/NuGet/NuGet.Client/pull/1073 was completed.

**Comment by tonyqus** [comment]
@zivkan Thank you for the clarification.

**Comment by tonyqus** [comment]
I created a new issue #12259  as you suggested. @zivkan
